### PR TITLE
osd: break multi condition asserts into multiple asserts

### PR DESCRIPTION
**Pull request by li3317**
Associated Trello: https://trello.com/c/Ri6ixQ5X/615-osd-cleanup-multi-condition-asserts
Signed-off-by: Annabel Li <li3317@purdue.edu>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/tools/rbd/action/MergeDiff.cc
Comment (idryomov) [inline src/tools/rbd/action/MergeDiff.cc:379]: This doesn't compile because `ceph_assert` is a macro and doesn't return anything.  If one really wanted to break this assert into two separate asserts, the equivalent would be:
```
ceph_assert(f_end);
ceph_assert(s_end);
```
But then I don't see the point of this change -- `f_end && s_end` is really not a complex condition.

**Review by idryomov — CHANGES_REQUESTED** [review]

File: src/tools/rbd/action/MergeDiff.cc
Comment (li3317) [reply src/tools/rbd/action/MergeDiff.cc:379]: Hi! this is a change from grace hopper open source day, i'm new to the project, the issue link is [here](https://trello.com/c/Ri6ixQ5X/615-osd-cleanup-multi-condition-asserts), do you think it'll make more sense to break multiple asserts in other more complex conditions? Thanks!

Comment (idryomov) [reply src/tools/rbd/action/MergeDiff.cc:379]: @tchaikov previously raised some concerns about `ceph_assert` not being completely free (potential executable size bloat, etc).  However the upside is better telemetry reports, so it's a bit of a balancing act.  I would suggest approaching these on a case by case basis:

- trello card mentions the OSD code specifically, this is mainly `src/osd` subdirectory
- pick a few of the really complex conditions there (`A && B` or `A || B` probably don't qualify unless `A` and/or `B` is itself complex)
- actually compile and test your change

Comment (ljflores) [reply src/tools/rbd/action/MergeDiff.cc:379]: @li3317 I know you had trouble setting up the project to compile-- let me check around in the OSD code for a better case.

Comment (ljflores) [reply src/tools/rbd/action/MergeDiff.cc:379]: Also @idryomov for context, the Open Source day participants had very limited time to set up the correct environment in order to compile. This is why @li3317 and a few others had to test their changes on make check instead of compiling locally.

Comment (ljflores) [reply src/tools/rbd/action/MergeDiff.cc:379]: @idryomov @li3317 this one could be worth breaking up, as the ceph_assert is already written on two lines:

https://github.com/ceph/ceph/blob/a79e029a727d594c67f1beb71ebeeb33f744583d/src/osd/ECBackend.cc#L2356-L2357

What do you guys think?

Comment (ljflores) [reply src/tools/rbd/action/MergeDiff.cc:379]: Also @li3317 if you need to go through with another change, I can provide you with the git commits you'd need to fix your PR. But first let's see what @idryomov thinks. :)

Comment (idryomov) [reply src/tools/rbd/action/MergeDiff.cc:379]: This one certainly seems less straightforward and therefore a better candidate, but this being OSD code it's up to @rzarzynski and @neha-ojha, not me ;)

Comment (ljflores) [reply src/tools/rbd/action/MergeDiff.cc:379]: @idryomov makes sense!

@li3317 if you're still interested in making the above change, let me know and I can help you with the git commands if needed.

Comment (li3317) [reply src/tools/rbd/action/MergeDiff.cc:379]: Hi @ljflores yes no problem I can make that change, and the git commands would be great thank you! I was finally able get the vm running but still have trouble setting some things up so I'm still not sure if I can compile locally though.

Comment (ljflores) [reply src/tools/rbd/action/MergeDiff.cc:379]: @li3317 no worries! If you have any issues setting up your vstart environment ( which is required to compile ) let me know and we can work through that too. For now, you can continue pushing new changes here to see whether it compiles on make check, and then fix it accordingly.

So what you'll need to do is:
1. Go to your terminal
2. `cd ceph`
3. `git checkout ghc-osd-li3317`
4. `git log`
5. Copy the SHA of the commit right *before* `tools/rbd/action: break multi condition asserts into multiple asserts` (the long line of letters and numbers right after the word "commit"

So if `git log` looks like this for you:
```
commit e78edb2b16672dfcc0ed648db0ded9a086d895e3 (HEAD -> ghc-osd-li3317, li3317/ghc-osd-li3317)
Author: li3317 <li3317@purdue.edu>
Date:   Fri Sep 16 15:45:30 2022 -0400

    tools/rbd/action: break multi condition asserts into multiple asserts
    
    Signed-off-by: Annabel Li <li3317@purdue.edu>

commit e4638a14ec4f27dd27f3e0707e7b03901e99813a (li3317/main)
Merge: ab52b8cc71c 0d6100bbf99
Author: Nizamudeen A <nia@redhat.com>
Date:   Fri Sep 16 20:49:48 2022 +0530

    Merge pull request #47849 from torchiaf/fix/37327
    
    mgr/dashboard: Add details to the modal which displays the `safe-to-d…
    
    Reviewed-by: Nizamudeen A <nia@redhat.com>
    Reviewed-by: Volker Theile <vtheile@suse.com>
```
You'll want to copy this SHA: `e4638a14ec4f27dd27f3e0707e7b03901e99813a`.

6. `git rebase -i <the SHA you copied`
7. Replace the word `pick` with `drop` since we want to get rid of this commit.
8. Save the commit (in VIM, this is just `:wq`.
9. Running `git log`, you can see that your branch is based on the main branch again, and we have gotten rid of the commit we don't want.

10. Next, you'll go to the file in https://github.com/ceph/ceph/pull/48150#discussion_r974752474 and make the change I described there.
11. Run `git diff` to make sure that your change looks as expected.
12. `git add <the file that you changed>`
13. `git commit`
14. This will bring you to a text file where you can write your commit. You can use the same structure as https://github.com/ceph/ceph/pull/48150/commits/e78edb2b16672dfcc0ed648db0ded9a086d895e3, except change the filepath from `tools/rbd/action` to `osd`, since that's where ECBackend.cc is located.
15. Don't forget to sign off your commit!
16. Save the commit (`:wq` on VIM)
17. Now, since we've re-written the history of the branch, we will have to force push the changes: `git push -f <the name of your forked repository> ghc-osd-li3317`
18. If you're unsure of the name of your forked repository, you can run `git remote -v` and see which name corresponds with https://github.com/li3317/ceph.git.

Then you're done! You've successfully updated your branch with the new commit.
When you're done, you can ping me (@ljflores) and let me know that it needs another review. And you can monitor make check (under the tests below) to see if your code compiles.

Comment (li3317) [reply src/tools/rbd/action/MergeDiff.cc]: Thank you so much @ljflores ! I've just pushed the change and I'll for sure save the list of git commands for future!

**Review by ljflores — APPROVED** [review]
Great, and it has compiled! (Note the green checkmark next to "make check").

**Comment by ljflores** [comment]
@neha-ojha @rzarzynski the idea behind this change is to improve debugging and help Telemetry group errors correctly.

**Comment by ljflores** [comment]
@li3317 now what will happen is that if this PR looks good to the above reviewers ^, I will add a "needs-qa" label, which means that the PR will go through additional integration testing. If you're interested, you can read more about that here: https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/tests-integration-testing-teuthology-intro/. If it passes the additional testing (we will update you on this), then the PR would be merged.

For now, nothing more is required on your end, but just keep an eye out for any other reviews or updates.

**Review by rzarzynski — APPROVED** [review]
LGTM!

**Comment by ljflores** [comment]
@neha-ojha adding "needs-qa", but let us know if you thing anything is amiss here.

**Review by neha-ojha — APPROVED** [review]
@li3317 It might be a good idea to add the motivation (as described in https://github.com/ceph/ceph/pull/48150#issuecomment-1255473714) behind this change to the commit message for future reference. Thanks for your contribution!

**Comment by li3317** [comment]
@ljflores done, thank you!

**Comment by ljflores** [comment]
@li3317 congratulations on getting your PR merged!

**Comment by ljflores** [comment]
[osd: cleanup multi-condition asserts](https://trello.com/c/Ri6ixQ5X/615-osd-cleanup-multi-condition-asserts)

**Pull request merged**
